HARK/distribution.py

```diff
@@ -930,29 +928,204 @@ def draw(self, N, X=None, exact_match=False):
+        """
+        Finds the distribution of a random variable Y that is a function
+        of discrete random variable X, Y=f(X).
+
+        Parameters
+        ----------
+        func : function
+            The function to be evaluated.
+            This function should take an array of shape dstn.dim().
+            It may also take other arguments *args.
+        *args :
+            Additional non-stochastic arguments for func,
+            The function is computed at f(dstn, *args).
+
+        Returns
+        -------
+        f_dstn : DiscreteDistribution
+            The distribution of func(dstn).
+        """
+        # we need to add one more dimension,
+        # the nature dimension, to any inputs that are n-dim arrays.
+        # This allows numpy to easily broadcast the function's output.
+        args = [
+            arg[..., np.newaxis] if isinstance(arg, np.ndarray) else arg for arg in args
+        ]
+        f_query = func(self.X, *args)
+
+        f_dstn = DiscreteDistribution(list(self.pmf), f_query, seed=self.seed)
+
+        return f_dstn
+
+
+class DiscreteDistributionXRA(DiscreteDistribution):
+    def __init__(
+        self,
+        pmf,
+        data,
+        seed=0,
+        coords=None,
+        dims=None,
+        name="DiscreteDistributionXRA",
+        attrs=None,
+    ):
+
+        if data.ndim < 2:
+            data = data[np.newaxis, ...]
+
+        if attrs is None:
+            attrs = {}
+
+        attrs["pmf"] = np.asarray(pmf)
+        attrs["seed"] = seed
+        attrs["RNG"] = np.random.RandomState(seed)
+
+        self._xarray = DataArray(
+            data=data,
+            coords=coords,
+            dims=dims,
+            name=name,
+            attrs=attrs,
+        )
+
+    @property
+    def xarray(self):
+        """
+        Returns the underlying xarray.DataArray object.
+        """
+        return self._xarray
+
+    @property
+    def X(self):
+        """
+        Returns the distribution's data as a numpy.ndarray.
+        """
+        return self._xarray.values
+
+    @property
+    def pmf(self):
+        """
+        Returns the distribution's probability mass function.
+        """
+        return self._xarray.pmf
+
+    @property
+    def RNG(self):
+        """
+        Returns the distribution's random number generator.
+        """
+        return self._xarray.RNG
+
+    @property
+    def data(self):
+        """
+        The distributions data as an array. The underlying
+        array type (e.g. dask, sparse, pint) is preserved.
+        """
+        return self._xarray.data
+
+    @property
+    def coords(self):
+        return self._xarray.coords
+
+    @property
+    def dims(self):
+        return self._xarray.dims
+
+    @property
+    def name(self):
+        return self._xarray.name
+
+    @property
+    def attrs(self):
+        return self._xarray.attrs
+
+    def calc_expectation(self, func=None, *args, labels=False):
```

Comment:
@sbenthall does this look good to you? Can it be more robust? I still get confused with how python interprets inputs when ordered instead of by key. 

Additionally, wondering if there is a cheap way to remove the boolean `labels`. I found `inspect.getsource(obj)` but I'm wondering if that'd be too much overhead. Haven't tried it.

Reply:
This looks great!  My only comment is on the ".X" point.  I have a strong preference for giving things names where there's a reasonable chance that one could do a global search-and-replace across the entire codebase and not mess everything up.  Replacing all occurences of the letter X would not satisfy that aim.  Elsewhere I think we have used "vals" and "prbs" for the corresponding objects.  Or, it could be "xvals" and "xprob" (I also like corresponding objects to have the same number of letters).  Maybe do a search through the codebase (and NARK) and see is what is most consistent with prior practice.

Reply:
Or, it could be ".mass" and ".prob"

Reply:
Will look into changing `.X` in a different PR.